### PR TITLE
chore: fixes for leanprover/lean4#2783

### PR DESCRIPTION
**Pull request by kim-em**
These are the adaptations required by leanprover/lean4#2783

Note that this PR is pointing at `nightly-testing-2023-11-06`, and cannot actually be merged. I will merge it into `nightly-testing` when it is needed.

A delegation or approving review would be helpful, so I know it is ready for the next Lean version bump!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
I have not finished making everything work, because the changes seem pretty sad to me.

@eric-wieser, is this what you were expecting from https://github.com/leanprover/lean4/issues/2042?

**Comment by eric-wieser** [comment]
After a quick skim, I think having a nicer way to unfold `Function.comp` would help here. We didn't need that in lean 3 because it was reducible.

One option would be to restore the "colon position determines unfolding" behavior we had in lean 3 (rather than I believe is the new "look for lambdas" heuristic), then define `comp f g x` with `x` after the colon.

Another would just be to add `comp_def` if it doesn't already exist, and use that throughout mathlib.

**Comment by semorrison** [comment]
Leo has also suggested that we could implement a `simp [!f]` syntax, that unfolds partial applications of `f`.

I'm hesitant to go that path, as *many* things interact with the `simp` syntax and surely there would be some annoying fallout with other tactics trying to reuse the simp syntax.

If we don't mind the additional complexity it is a concise solution at the point of use.

`comp_def` is probably also fine. I guess tomorrow I will continue fixing up this PR.

**Comment by eric-wieser** [comment]
My guess would be that in almost all cases in mathlib, it's the case that if we are relying on `simp` unfolding `f` and it's not doing so, this is because we *always* want `f` to unfold in this way, not because we want it to do so at this particular call site. So I think our options are:
* Add new lemmas for these cases
* Register custom equation lemmas for these cases uses `@[eqns]`
* Have a way to indicate what unfolding is desired in the definition; perhaps if the `:`-detection is considered ugly, it could be something like an `fun x => unfold_unapplied (fun y => ...)` syntax, where `unfold_unapplied` is metadata like `no_index`, and tells `simp` that lambdas inside it do not need to be applied in order to allow unfolding.

**Comment by semorrison** [comment]
@eric-wieser I think my preference would be to go with option 1 or 2 on your list, with the alternative of revisiting option 3 (but via a new RFC / PR, rather than changing leanprover/lean4#2783).

That would mean:
* finish adding the workarounds as in the PR
* merge leanprover/lean4#2783
* at leisure, Mathlib adds the additional `simp` or `eqns` annotations and cleans up the workarounds?

File: scripts/count-transitive-imports.lean
Comment (eric-wieser) [inline scripts/count-transitive-imports.lean:1]: Is this supposed to be here?

**Comment by eric-wieser** [comment]
I just pushed 95ee614e2aa40e19db886617a3cdc56ca25b1b0d but it's not showing up...

File: scripts/count-transitive-imports.lean
Comment (eric-wieser) [reply scripts/count-transitive-imports.lean:1]: Removed

**Pull request closed without merging**